feat: add `monitor` to operator, fix monitoring setup

mjnagel opened the pull request:
## Description

This adds a new portion to the `Package` spec, `monitor`. This functionality is briefly documented [here](https://github.com/defenseunicorns/uds-core/blob/operator-monitor-magic/docs/MONITOR.md). Also included is a mutation for all service monitors to handle them in the expected istio mTLS way. This PR also fixes some missing dashboards and enables all default dashboards from the upstream kube-prometheus-stack chart (this may be excessive?).

Currently monitored:
- Prometheus stack (operator, self, alertmanager)
- Loki
- kube-system things (kubelet, coredns, apiserver)
- Promtail
- Metrics-server
- Velero
- Keycloak
- Grafana
- All istio envoy proxies (podmonitor)

Not added here:
- NeuVector: Currently has limited config options with regards to auth, keeping this disabled in anticipation of SSO on NeuVector, with a desire to contribute upstream to enable our use case of SSO-only auth.

In addition this PR switches single package testing to use/add Istio. This appears to add around 1 minute to each pipeline run, but:
- allows us to test the istio VS endpoints in single package checks (some quirks with this in current state due to SSO redirects, but will allow us to do e2e testing in those pipelines in the future)
- allows us to assume istio always and not build bespoke pepr code for the specific no-istio scenario (ex: prometheus can assume certs)

This would still allow someone to run locally without istio for some packages / un-inject, mess around with mTLS, etc - which may be useful still to identify where Istio is causing problems. This just switches our CI posture so that we assume istio always.

## Related Issue

Fixes https://github.com/defenseunicorns/uds-core/issues/17

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Other (security config, docs update, etc)

## Checklist before merging

- [x] Test, docs, adr added or updated as needed
- [x] [Contributor Guide Steps](https://github.com/defenseunicorns/uds-template-capability/blob/main/CONTRIBUTING.md)(https://github.com/defenseunicorns/uds-template-capability/blob/main/CONTRIBUTING.md#submitting-a-pull-request) followed

mjnagel committed:
wip: servicemonitor generation

mjnagel committed:
wip: more things

mjnagel committed:
wip: rebase

mjnagel committed:
wip: doc

mjnagel committed:
wip

mjnagel committed:
wip: docs, refactor

mjnagel committed:
wip: cleanup

mjnagel committed:
wip: prometheus monitor

mjnagel committed:
wip: handle timing

mjnagel committed:
Merge branch 'main' into operator-monitor-magic

mjnagel committed:
Merge branch 'main' into operator-monitor-magic

mjnagel committed:
wip: docs + neuvector

mjnagel committed:
Merge branch 'main' into operator-monitor-magic

mjnagel committed:
wip: rebase main [ci skip]

mjnagel committed:
wip: fix thingy [ci skip]

mjnagel committed:
Merge branch 'main' into operator-monitor-magic

mjnagel committed:
wip: dashboards, search all, noprogess [ci skip]

mjnagel committed:
Merge branch 'main' into operator-monitor-magic

mjnagel committed:
[skip ci]

mjnagel committed:
wip: rebase [skip ci]

mjnagel committed:
wip: docs [ci skip]

mjnagel committed:
Merge branch 'main' into operator-monitor-magic

mjnagel committed:
wip: rebase, neuvector doesn't work [ci skip]

mjnagel committed:
[ci skip]
Merge branch 'main' into operator-monitor-magic

mjnagel committed:
Merge branch 'main' into operator-monitor-magic

mjnagel committed:
wip: add velero/metrics-server

mjnagel committed:
wip: grafana metrics

mjnagel committed:
Merge branch 'main' into operator-monitor-magic

mjnagel committed:
wip: add keycloak

mjnagel committed:
fix: bad yaml, docs: monitoring doc

mjnagel committed:
Merge branch 'main' into operator-monitor-magic

mjnagel committed:
chore: pepr cleanup

mjnagel committed:
fix: alertmanager tls config

mjnagel committed:
wip: failing single package comment [ci skip]

mjnagel commented on src/prometheus-stack/values/values.yaml:
```diff
@@ -132,3 +94,13 @@ prometheusOperator:
+      caFile: /etc/prom-certs/root-cert.pem
+      certFile: /etc/prom-certs/cert-chain.pem
+      insecureSkipVerify: true
+      keyFile: /etc/prom-certs/key.pem
```
todo: fix this for single package testing...

TristanHoladay committed:
Merge branch 'main' into operator-monitor-magic

TristanHoladay commented:
neuvector chart `2.7.4` was released with your changes @mjnagel . https://github.com/defenseunicorns/uds-core/pull/333 tracking the upgrade. waiting for IB to update images to `5.3.1`

mjnagel committed:
Merge branch 'main' into operator-monitor-magic

mjnagel committed:
no-progress [skip ci]

renovate[bot] committed:
chore(deps): update neuvector

| datasource | package                                                            | from  | to    |
| ---------- | ------------------------------------------------------------------ | ----- | ----- |
| helm       | core                                                               | 2.7.3 | 2.7.6 |
| helm       | crd                                                                | 2.7.3 | 2.7.6 |
| docker     | docker.io/neuvector/controller                                     | 5.3.0 | 5.3.2 |
| docker     | docker.io/neuvector/enforcer                                       | 5.3.0 | 5.3.2 |
| docker     | docker.io/neuvector/manager                                        | 5.3.0 | 5.3.2 |
| docker     | docker.io/neuvector/prometheus-exporter                            | 5.3.0 | 5.3.2 |
| docker     | ironbank/neuvector/neuvector/prometheus-exporter                   | 5.3.0 | 5.3.2 |
| helm       | monitor                                                            | 2.7.3 | 2.7.6 |
| docker     | neuvector/prometheus-exporter                                      | 5.3.0 | 5.3.2 |
| docker     | registry1.dso.mil/ironbank/neuvector/neuvector/controller          | 5.3.0 | 5.3.2 |
| docker     | registry1.dso.mil/ironbank/neuvector/neuvector/enforcer            | 5.3.0 | 5.3.2 |
| docker     | registry1.dso.mil/ironbank/neuvector/neuvector/manager             | 5.3.0 | 5.3.2 |
| docker     | registry1.dso.mil/ironbank/neuvector/neuvector/prometheus-exporter | 5.3.0 | 5.3.2 |

mjnagel committed:
chore: string silliness

mjnagel committed:
chore: renovate update for strings

mjnagel committed:
Merge branch 'main' into operator-monitor-magic

mjnagel committed:
wip: fix neuvector

mjnagel committed:
Merge branch 'main' into operator-monitor-magic

mjnagel committed:
comment

mjnagel committed:
ci: [skip ci]

mjnagel committed:
wip: add istio to single pkg test for time check

mjnagel committed:
Merge branch 'main' into operator-monitor-magic

mjnagel committed:
wip: get some time metrics

mjnagel committed:
Merge branch 'main' into operator-monitor-magic

mjnagel committed:
ci: delete delete delete

mjnagel committed:
docs: cleanup comment

mjnagel committed:
fix: linting, interface checks

mjnagel committed:
302 for grafana

mjnagel committed:
beautiful grafana curling

mjnagel committed:
posix land

mjnagel commented on src/prometheus-stack/values/values.yaml:
```diff
@@ -35,6 +19,8 @@ nodeExporter:
   serviceMonitor:
     interval: ""
 prometheus:
```
Add istio pod monitor?

mjnagel committed:
fix: add istio podmonitor for envoy stats

mjnagel committed:
Merge branch 'main' into operator-monitor-magic

mjnagel committed:
Merge branch 'main' into operator-monitor-magic

mjnagel committed:
Merge branch 'main' into operator-monitor-magic

mjnagel committed:
Merge branch 'main' into operator-monitor-magic

rjferguson21 commented on src/pepr/prometheus/index.ts:
```diff
@@ -0,0 +1,60 @@
+  .IsCreatedOrUpdated()
+  .Mutate(async sm => {
+    // Provide an opt-out of mutation to handle complicated scenarios
+    if (sm.Raw.metadata?.labels?.["uds/skip-sm-mutate"]) {
```
Would an annotation be a better fit here?

rjferguson21 commented on src/pepr/operator/controllers/monitoring/service-monitor.ts:
```diff
@@ -0,0 +1,86 @@
+      },
+      spec: {
+        endpoints: endpoints,
+        selector: promSelector,
```
nit: I'd probably just inline these variables.

rjferguson21 commented on src/pepr/operator/controllers/monitoring/service-monitor.ts:
```diff
@@ -0,0 +1,86 @@
+  }
+
+  // Return the list of monitor names
+  return [...new Set(payloads.map(sm => sm.metadata!.name!).flat())];
```
Could this just be the `payloads.map`?

rjferguson21 commented on src/pepr/operator/controllers/network/policies.ts:
```diff
@@ -59,6 +59,28 @@ export async function networkPolicies(pkg: UDSPackage, namespace: string) {
+    const policy: Allow = {
+      direction: Direction.Ingress,
+      selector: podSelector ?? selector,
+      remoteNamespace: `monitoring`,
```
```suggestion
      remoteNamespace: "monitoring",
```

rjferguson21 commented on src/pepr/operator/controllers/network/policies.ts:
```diff
@@ -59,6 +59,28 @@ export async function networkPolicies(pkg: UDSPackage, namespace: string) {
+      selector: podSelector ?? selector,
+      remoteNamespace: `monitoring`,
+      remoteSelector: {
+        app: `prometheus`,
```
```suggestion
        app: "prometheus",
```

rjferguson21 commented on src/pepr/prometheus/index.ts:
```diff
@@ -0,0 +1,60 @@
+        namespace.metadata.labels["istio-injection"] === "enabled"
+      ) {
+        istioInjected = true;
+      }
```
I think this could be pulled out into a helper function like `isIstioInjected`.  I believe you are after if any of the namespaces (or namespace of the SM) are istio injected.

rjferguson21 commented on src/pepr/operator/controllers/monitoring/service-monitor.ts:
```diff
@@ -0,0 +1,86 @@
+    };
+
+    // Apply the VirtualService and force overwrite any existing policy
+    await K8s(Prometheus.ServiceMonitor).Apply(payload, { force: true });
```
Not going to hold up this PR but if we could do our best to avoid mixing the code for "resource building" and us using KFC to apply it will make unit testing this code a lot easier.  

It is a small thing but wherever we can segregate pure functions we should try to.

mjnagel replied:
Lmk if this fits what you expected - https://github.com/defenseunicorns/uds-core/pull/256/commits/9cbd7db3cc8334930761c52b6c52c631fe79c86d - unit test is a bit eh, but attempted to just pull that generation piece out. Think I could've recreated/grabbed some of the vars in the function itself rather than passing them, but they're needed in both spots so it seemed 50/50.

mjnagel committed:
Merge branch 'main' into operator-monitor-magic

mjnagel committed:
wip: pr feedback [skip ci]

mjnagel committed:
pull out function, mini unit test

mjnagel committed:
format

mjnagel committed:
delete delete delete
